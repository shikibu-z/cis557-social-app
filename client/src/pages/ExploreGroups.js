/* eslint-disable consistent-return */
/* eslint-disable prefer-template */
import React, { useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import {
  Dropdown, DropdownButton, Button, Toast, Stack, Form, Row, Col,
} from 'react-bootstrap';
import NavBar from '../components/NavBar';
import HomePageGroup from '../components/HomePageGroup';
import '../stylesheets/ExploreGroups.css';

import {
  getTopics, getPublicGroups, getSortedGroups, getSuggestedGroups,
} from '../api/fetchers';

function ExploreGroups() {
  // groups after name search & topic filter
  const [groups, setGroups] = useState([]);
  // groups fetched from remote by tags
  const [fetchedGroups, setFetchedGroups] = useState([]);
  // the group name query user input
  const [groupQuery, setGroupQuery] = useState('');

  // user input topic
  const [topic, setTopic] = useState('');
  // group tag filter
  const [filterTopics, setFilterTopics] = useState([]);
  const [popularTopics, setPopularTopics] = useState([]);

  const [showFilter, setShowFilter] = useState(false);

  const history = useHistory();
  const id = localStorage.getItem('currUser');

  // get popular topics if there's one when loading the page
  const updatePopularTopics = async () => {
    try {
      const tmp = [];
      const topics = await getTopics();
      if (topics.data.length !== 0) {
        topics.data.forEach((entry) => {
          tmp.push(entry.topic);
        });
      }
      const response = await getPublicGroups(id);
      setPopularTopics(tmp);
      setGroups(response.data);
      setFetchedGroups(response.data);
    } catch (err) {
      return err;
    }
  };
  useEffect(() => {
    updatePopularTopics();
  }, []);

  // fetch groups by provided topics
  const fetchRemoteByTopics = async () => {
    try {
      let query = id;
      const tmp = [];
      if (filterTopics.length !== 0) {
        filterTopics.forEach((entry) => {
          tmp.push(entry + '=true');
        });
        query += '?' + tmp.join('&');
      }
      const response = await getPublicGroups(query);
      setGroups(response.data);
      setFetchedGroups(response.data);
    } catch (err) {
      return err;
    }
  };
  useEffect(() => {
    fetchRemoteByTopics();
  }, [filterTopics]);

  // filter fetched groups by name
  const handleFilterGroups = (e) => {
    const query = e.target.value;
    if (query !== '') {
      const res = groups.filter(
        (group) => group.group.groupName.toLowerCase().startsWith(query.toLowerCase()),
      );
      setGroups(res);
    } else {
      setGroups(fetchedGroups);
    }
    setGroupQuery(query);
  };

  // remove a topic tag
  const removeTag = (e) => {
    const toRemove = e.target.innerText;
    const newArr = filterTopics.filter((item) => item !== toRemove);
    setFilterTopics(newArr);
  };

  // read manual input tag
  const handleChange = (e) => {
    setTopic(e.target.value);
  };

  // add manual input tag to filter list
  const addTopic = () => {
    if (!filterTopics.includes(topic.toLowerCase())) {
      setFilterTopics([...filterTopics, topic.toLowerCase()]);
    }
    setTopic('');
  };

  // add provided popular tags to filter list
  const handleAddTopic = (e) => {
    const toAdd = e.target.innerText;
    if (!filterTopics.includes(toAdd.toLowerCase())) {
      setFilterTopics([...filterTopics, toAdd.toLowerCase()]);
    }
  };

  const handleSorting = async (e) => {
    try {
      let query = e.toString();
      const tmp = [];
      if (filterTopics.length !== 0) {
        filterTopics.forEach((entry) => {
          tmp.push(entry + '=true');
        });
        query += '?' + tmp.join('&');
      }
      const response = await getSortedGroups(query);
      setFetchedGroups(response.data);
      setGroups(response.data);
    } catch (err) {
      return err;
    }
  };

  const handleSuggested = async (event) => {
    try {
      if (event === 'all') {
        await updatePopularTopics();
      } else if (event === 'suggested') {
        const response = await getSuggestedGroups(id);
        setFetchedGroups(response.data);
        setGroups(response.data);
      }
    } catch (err) {
      return err;
    }
  };

  return (
    <div>
      <NavBar isLogin={0} />
      <div className="home">
        <div className="groupSide d-flex">
          <div className="exploregroup-header">
            <input id="group-search" value={groupQuery} className="homepage-group-search" name="group-search" placeholder="Search a group" onChange={handleFilterGroups} />
            <DropdownButton onSelect={handleSorting} variant="secondary" title="Sort By">
              <Dropdown.Item eventKey="recent" as="button">Most Recent</Dropdown.Item>
              <Dropdown.Item eventKey="post" as="button">Most Posts</Dropdown.Item>
              <Dropdown.Item eventKey="member" as="button">Most Members</Dropdown.Item>
            </DropdownButton>
            <Button style={{ 'background-color': '#C4C4C4', 'border-color': '#C4C4C4' }} variant="secondary" onClick={() => { setShowFilter(true); }}>Filter By</Button>
            <Toast
              onClose={() => setShowFilter(false)}
              show={showFilter}
              delay={3000}
              style={{
                position: 'absolute', top: '27%', left: '55%', width: '500px',
              }}
            >
              <Toast.Header closeButton />
              <Toast.Body>
                <Stack direction="vertical" gap={3}>
                  <div>
                    <big>Filtered Topic</big>
                    <Form>
                      <Form.Group as={Row} className="justify-content-md-center">
                        <Col md={6}>
                          <Form.Control placeholder="Topic" value={topic} onChange={handleChange} />
                        </Col>
                        <Col md={2}>
                          <Button onClick={addTopic}>Add</Button>
                        </Col>
                      </Form.Group>
                    </Form>
                  </div>
                  <div>
                    <big>PopularTopics</big>
                    <div>
                      {popularTopics.map((tag) => (
                        <Button key={tag} onClick={handleAddTopic}>{tag}</Button>
                      ))}
                    </div>
                  </div>
                </Stack>
              </Toast.Body>
            </Toast>
            <DropdownButton onSelect={handleSuggested} variant="secondary" title="Suggested">
              <Dropdown.Item eventKey="all" as="button">All Groups</Dropdown.Item>
              <Dropdown.Item eventKey="suggested" as="button">Suggested Groups</Dropdown.Item>
            </DropdownButton>
            <Button id="btn-creategroup" size="lg" onClick={() => { history.push('/creategroup'); }}>Create Group</Button>
          </div>
          <div className="mt-2 col-md-12" />
          <div>
            <h3 style={{ 'margin-left': '1%' }}>Filtered Topics:</h3>
            <div style={{ 'margin-left': '1%' }}>
              {filterTopics && (
                filterTopics.map((tag) => (
                  <Button variant="dark" key={tag} onClick={removeTag}>
                    {tag}
                  </Button>
                ))
              )}
            </div>
          </div>
          <div className="mt-2 col-md-12" />
          <div id="homepage-groups">
            {groups.map((g) => (
              <HomePageGroup
                id={g.group.idGroups}
                src={g.group.profilePhoto}
                groupName={g.group.groupName}
                groupInfo={g.group.groupInfo}
                members={g.numOfMembers}
                priv={0}
                join={g.isJoined.toString()}
              />
            ))}
          </div>
        </div>
      </div>
    </div>
  );
}

export default ExploreGroups;
