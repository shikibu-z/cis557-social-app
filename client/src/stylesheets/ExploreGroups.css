.exploregroup-header {
  height: 10%;
  display: flex;
  flex-direction: row;
  align-items:center;
  text-align: center;
  margin-left: 3%;
  margin-top: 1%;
  gap: 10px;
}
.exploregroup-header > #btn-creategroup {
  margin-left: auto;
  margin-right: 3%;
}

.dropdown-toggle {
  background-color: #D9D9D9;
  border-color:#D9D9D9; 
}