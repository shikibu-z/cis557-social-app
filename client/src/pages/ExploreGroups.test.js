/*eslint-disable */
import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import ExploreGroups from './ExploreGroups';

test('Exploregroup page matches snapshot', () => {
  const component = renderer.create(<BrowserRouter><ExploreGroups /></BrowserRouter>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
