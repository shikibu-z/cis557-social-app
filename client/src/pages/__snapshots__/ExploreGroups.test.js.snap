// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exploregroup page matches snapshot 1`] = `
<div>
  <div>
    <nav
      className="menu navbar navbar-expand navbar-dark bg-dark"
    >
      <div
        className="container-fluid"
      >
        <a
          className="navbar-brand"
          href="#home"
        >
          <img
            alt="connexion"
            className="d-inline-block align-top"
            id="logo"
            src="../Connexion-logos_transparent.png"
          />
           
        </a>
        <div>
          <button
            aria-controls="navbarScroll"
            aria-label="Toggle navigation"
            className="navbar-toggler collapsed"
            onClick={[Function]}
            type="button"
          >
            <span
              className="navbar-toggler-icon"
            />
          </button>
          <div
            aria-expanded={null}
            className="navbar-collapse collapse"
            id="navbarScroll"
          >
            <div
              className="me-auto my-2 my-lg-0 navbar-nav navbar-nav-scroll"
              onKeyDown={[Function]}
              style={
                Object {
                  "maxHeight": "100px",
                }
              }
            >
              <button
                className="NavBar-buttons btn btn-primary"
                disabled={false}
                id="home"
                onClick={[Function]}
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6.5 14.5v-3.505c0-.245.25-.495.5-.495h2c.25 0 .5.25.5.5v3.5a.5.5 0 0 0 .5.5h4a.5.5 0 0 0 .5-.5v-7a.5.5 0 0 0-.146-.354L13 5.793V2.5a.5.5 0 0 0-.5-.5h-1a.5.5 0 0 0-.5.5v1.293L8.354 1.146a.5.5 0 0 0-.708 0l-6 6A.5.5 0 0 0 1.5 7.5v7a.5.5 0 0 0 .5.5h4a.5.5 0 0 0 .5-.5z"
                  />
                </svg>
              </button>
              <button
                className="NavBar-buttons btn btn-primary"
                disabled={false}
                id="createGroup"
                onClick={[Function]}
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 13a1.5 1.5 0 0 0 1.5 1.5h13A1.5 1.5 0 0 0 16 13V6a1.5 1.5 0 0 0-1.5-1.5h-13A1.5 1.5 0 0 0 0 6v7zM2 3a.5.5 0 0 0 .5.5h11a.5.5 0 0 0 0-1h-11A.5.5 0 0 0 2 3zm2-2a.5.5 0 0 0 .5.5h7a.5.5 0 0 0 0-1h-7A.5.5 0 0 0 4 1z"
                  />
                </svg>
              </button>
              <button
                className="NavBar-buttons btn btn-primary"
                disabled={false}
                id="chat"
                onClick={[Function]}
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 15c4.418 0 8-3.134 8-7s-3.582-7-8-7-8 3.134-8 7c0 1.76.743 3.37 1.97 4.6-.097 1.016-.417 2.13-.771 2.966-.079.186.074.394.273.362 2.256-.37 3.597-.938 4.18-1.234A9.06 9.06 0 0 0 8 15z"
                  />
                </svg>
              </button>
              <button
                className="NavBar-buttons btn btn-primary"
                disabled={false}
                id="profile"
                onClick={[Function]}
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3 14s-1 0-1-1 1-4 6-4 6 3 6 4-1 1-1 1H3zm5-6a3 3 0 1 0 0-6 3 3 0 0 0 0 6z"
                  />
                </svg>
              </button>
              <button
                className="NavBar-buttons btn btn-primary"
                disabled={false}
                id="logout"
                onClick={[Function]}
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6 3.5a.5.5 0 0 1 .5-.5h8a.5.5 0 0 1 .5.5v9a.5.5 0 0 1-.5.5h-8a.5.5 0 0 1-.5-.5v-2a.5.5 0 0 0-1 0v2A1.5 1.5 0 0 0 6.5 14h8a1.5 1.5 0 0 0 1.5-1.5v-9A1.5 1.5 0 0 0 14.5 2h-8A1.5 1.5 0 0 0 5 3.5v2a.5.5 0 0 0 1 0v-2z"
                    fillRule="evenodd"
                  />
                  <path
                    d="M11.854 8.354a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H1.5a.5.5 0 0 0 0 1h8.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3z"
                    fillRule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </div>
  <div
    className="home"
  >
    <div
      className="groupSide d-flex"
    >
      <div
        className="exploregroup-header"
      >
        <input
          className="homepage-group-search"
          id="group-search"
          name="group-search"
          onChange={[Function]}
          placeholder="Search a group"
          value=""
        />
        <div
          className="dropdown"
        >
          <button
            aria-expanded={false}
            className="dropdown-toggle btn btn-secondary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Sort By
          </button>
        </div>
        <button
          className="btn btn-secondary"
          disabled={false}
          onClick={[Function]}
          style={
            Object {
              "background-color": "#C4C4C4",
              "border-color": "#C4C4C4",
            }
          }
          type="button"
        >
          Filter By
        </button>
        <div
          className="dropdown"
        >
          <button
            aria-expanded={false}
            className="dropdown-toggle btn btn-secondary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Suggested
          </button>
        </div>
        <button
          className="btn btn-primary btn-lg"
          disabled={false}
          id="btn-creategroup"
          onClick={[Function]}
          type="button"
        >
          Create Group
        </button>
      </div>
      <div
        className="mt-2 col-md-12"
      />
      <div>
        <h3
          style={
            Object {
              "margin-left": "1%",
            }
          }
        >
          Filtered Topics:
        </h3>
        <div
          style={
            Object {
              "margin-left": "1%",
            }
          }
        />
      </div>
      <div
        className="mt-2 col-md-12"
      />
      <div
        id="homepage-groups"
      />
    </div>
  </div>
</div>
`;
